### PR TITLE
Contribute DebugCounter documentation

### PR DESCRIPTION
**Pull request by amankumar1**
Contribute compiler/ras/DebugCounter.md.

Documenation constains information about using debug counters,
different types of debug counters, naming schemes, and
specifying output.

Signed-off-by: Aman Kumar amank@ca.ibm.com

### CODE REVIEW

File: compiler/ras/DebugCounter.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
+# Debug Counter
+
+A debug counter is a facility for adding instrumentation to jitted code to count arbitrary things.
```
Comment (andrewcraik) [inline compiler/ras/DebugCounter.md]: There is also an incStaticDebugCounter which can be used to count things at compile-time. Perhaps a better introduction would be to say debug counters are a general infrastructure for counting things. There are two flavours - static and dynamic. Static counters do not add anything to the code generated by the compiler, but server to count events at compile-time while dynamic counters are incremented at runtime and are used to count events in code generated by the compiler.

File: compiler/ras/DebugCounter.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+Counters below a certain fidelity can be excluded by using the `-Xjit:debugCounterFidelity=##` option.
+
+## Static and Dynamic Counters
+
+Whenever you use a counter using either `TR_Compilation::prependDebugCounter` or `TR::CodeGenerator::generateDebugCounter`, two counters are actually created: one static, one dynamic. Dynamic counters are implemented by injecting nodes into trees or instructions into the generated code, and they count how often the particular code path is reached at runtime. Static counters are incremented at compilation, when the `prependDebugCounter` or `generateDebugCounter` method is called. The delta used for each counter can be specified separately using the available overloads, though they are the same by default.
+
+## Naming Scheme
+
+The name of a counter can affect its behaviour, and its relationship with other counters. Special characters in a counter name signal the debug counter facility to treat a counter in a special way.
```
Comment (andrewcraik) [inline compiler/ras/DebugCounter.md]: The name does not affect the counter behaviour - it modifies the display processing logic for counters which affects how the results are shown to the user - the raw counter will continue to bump in the same way.

File: compiler/ras/DebugCounter.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+#### `staticDebugCounters=`
+
+Specifies a [regular expression](../../doc/compiler/regular_expression/Regular_Expression.md) that is matched against static debug counter names to determine which are included in the output.
+
+### Example
+
+```
+export TR_DebugCounterFileName=debugcounters.txt
+```
+Setting the option to `debugCounters={*}` will output all dynamic debug counters.
```
Comment (andrewcraik) [inline compiler/ras/DebugCounter.md]: This is not a useful thing in general due to the nature of debug counters and is probably not worth documenting.

**Comment by amankumar1** [comment]
@andrewcraik Documentation updated.

**Comment by andrewcraik** [comment]
Thanks for accommodating those changes. To round out the documentation it also should discuss incStaticDebugCounter which creates a static debug counter without a dynamic counterpart. These are used quite a bit in the optimizer to count compile events that we do not need or cannot count at runtime. That is to say static only counters can and do exist, dynamic counters get static counterparts by default since knowing the ratio of the number of counters inserted vs the number of counter bumps observed at runtime is a generally useful thing for a dynamic counter.

**Comment by vijaysun-omr** [comment]
Do you think having example usage(s) shown in the documentation would be useful ? It might serve as a quick reminder on what the syntax is, rather than have to read the full documentation.

**Comment by andrewcraik** [comment]
Yes - perhaps having a 'Quick Summary' section with example syntax and the full documentation below might help and addition to the incStaticDebugCounter notes.

**Comment by amankumar1** [comment]
@vijaysun-omr @andrewcraik Added example syntax and made some other minor changes and additions.

**Comment by vijaysun-omr** [comment]
One more small request : please added the -Xjit command lines one would need to use for the examples you showed. Thanks.

**Comment by amankumar1** [comment]
@vijaysun-omr Added the command line options for the examples.

**Comment by vijaysun-omr** [comment]
Looks good. Thanks Aman.

**Pull request merged**